docs/proposals/20230718-inplace-workload-vertical-scaling.md

```diff
@@ -0,0 +1,813 @@
++                       }
++               }
+
+-func getRequestedResources(container *v1.Container) (map[v1.ResourceName]uint64, error) {
++func getRequestedResources(pod *v1.Pod, container *v1.Container) (map[v1.ResourceName]uint64, error) {
+        requestedResources := map[v1.ResourceName]uint64{}
+-       for resourceName, quantity := range container.Resources.Requests {
++       resources := container.Resources.Requests
++       if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
++               if cs, ok := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name); ok {
++                       resources = cs.ResourcesAllocated
++               }
++       }
++       for resourceName, quantity := range resources {
+                if resourceName != v1.ResourceMemory && !corehelper.IsHugePageResourceName(resourceName) {
+                        continue
+                }
+```
+
+Added an interface to refresh the pod cache in pleg, once the resize state is InProgress (Pod resource changes can take effect at any time), At the end of each syncPod round, the UpdateCache method calls getSandboxIDByPodUID at kuberuntime to get the latest pod information.
+
+```go
+@@ -1842,6 +1858,15 @@ func (kl *Kubelet) syncPod(_ context.Context, updateType kubetypes.SyncPodType,
+                return false, nil
+        }
+
++       if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && isPodResizeInProgress(pod, &apiPodStatus) {
++               // While resize is in progress, periodically call PLEG to update pod cache
++               runningPod := kubecontainer.ConvertPodStatusToRunningPod(kl.getRuntime().Type(), podStatus)
++               if err := kl.pleg.UpdateCache(&runningPod, pod.UID); err != nil {
++                       klog.ErrorS(err, "Failed to update pod cache", "pod", klog.KObj(pod))
++                       return false, err
++               }
++       }
++
+        return false, nil
+ }
+```
+
+The corresponding changes to validation are not expanded here.
+
+### Goals
+
+- Users only need to modify template in workload to achieve vertical scaling for all managed Pods
+- Simple and clear configuration strategy, support grayscale capabilities
+- Provides the rollback function, and attempts to return to the previous state when the capacity expansion and shrinkage fails. If the rollback still fails, no further action is required
+
+### Non-Goals/Future Work
+
+- Implement preemption ensures that all Pods can successfully expand and shrink
+
+## Proposal
+
+### User Stories
+
+Users can start vertical scaling of all managed Pods by modifying the resource field in workload.
+
+### Implementation Details/Notes/Constraints
+
+#### Design Objectives
+
+- Use vscale as a standalone operation rather than integrated in update. For example, for cloneset, three operations will be improved to scale, vscale, and update.
```

Comment:
Good idea! But we should consider how to avoid to break `maxUnavailable`  limitation in some scene, for example, users update both resources and images fields, and settings `RestartRequired` of vscale.